keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,404 @@
+
+### User Stories 
+
+#### Story 1
+
+A development environment implemented as a Kubernetes Pod allows for separation of tools and a (web-based) IDE between multiple side-cars.
+
+The development environment defines a container with the web-server serving the IDE itself, and constrains it to use a certain amount of memory. Additional tools are provided in additional side-car deployments - for example an [LSP](https://langserver.org/) service, a terminal, and more. 
+
+Using this new feature, the containers providing the terminal, the LSP services, and the set of tools being utilized can share the resource limit defined for the pod. Consider the following pod definition:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: ide
+  name: myide
+spec:
+  resources:
+    limits:
+      memory: "1024M"
+      cpu: "4"
+  containers:
+  - image: shell
+    name: debian:buster
+  - image: tool1
+    name: first-tool:latest
+  - image: tool2
+    name: second-tool:latest
+  - name: ide
+    image: theia:latest
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+Using the pod-level resource definition enables the `tool1` and `tool2` containers to be constrained by the total limit for the pod. 
+
+Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.
+
+### Implementation Details/Notes/Constraints 
+
+The proposal is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should turn it on by specifying the relevant resource section in the Pod specification.
+
+### Interactions With Other Features
+
+#### Support for cgroup v2
+
+This proposal is compatible with both cgroup v1 and cgroup v2. Both versions of cgroup allow specifying limits on all levels of the cgroup hierarchy.
+
+#### ResourceOverhead
+
+Since the runtime shim is started in the pod-level cgroup, there should be no bad interactions between this proposed feature and the ResourceOverhead feature.
+
+#### Memory-based Emptydir volumes
+
+This proposal doesn't change the current status quo in any way. Memory used for files in a `tmpfs` volume is accounted for as shared-memory. The Linux kernel charges the used memory to cgroup that first touched the shared page for the memory. Moving the charge for shared memory pages between cgroups is not currently supported by the Linux kernel. Defining a memory limit on the pod cgroup level would not change the underlying limitation that memory used by files in the volume are accounted to the cgroup that first touched the memory.
+
+#### HugeTLB cgroup
+
+In progress
```

Comment:
Any updates on this, or is it supposed to be a non-goal to support?

I have some insight into hugetlb, and it will work more or less the same way as normal memory will when using pod-level limits. In contrast to memory where unset means `unlimited`, when hugetlb is unset it is essentially the same as enforcing `0`. This means that when a pod-level request/limit is set for a given hugetlb size, and some containers haven't set any limit, the pod cgroup should probably be set to `unlimited` or the same value as the pod-level cgroup.

If that makes sense to you?

Reply:
Makes sense. 
So basically: 
1. If no pod-level limits are specified for `hugetlb`, make no change to the current implementation.
1. If pod-level limits are set, apply them at the pod-level cgroup.
    1. for each container in the pod, 
         1. if container-level limits are set, use the values as provided
         1. if container-level limits are NOT set, then set the container-level cgroup to unlimited.